pkg/admission/customresourcedefinition/admission.go

```diff
@@ -0,0 +1,85 @@
+/*
+Copyright 2022 The KCP Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package customresourcedefinition
+
+import (
+	"context"
+	"fmt"
+	"io"
+
+	"github.com/kcp-dev/apimachinery/pkg/logicalcluster"
+
+	"k8s.io/apiextensions-apiserver/pkg/apis/apiextensions"
+	apiextensionsv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	"k8s.io/apiserver/pkg/admission"
+	"k8s.io/apiserver/pkg/endpoints/request"
+
+	"github.com/kcp-dev/kcp/pkg/reconciler/apis/apibinding"
+)
+
+const (
+	PluginName            = "apis.kcp.dev/ReservedBoundCRDAnnotation"
+	annotationKeyBoundCRD = "apis.kcp.dev/bound-crd"
+)
+
+func Register(plugins *admission.Plugins) {
+	plugins.Register(PluginName,
+		func(_ io.Reader) (admission.Interface, error) {
+			return &customResourceDefinitionBound{
+				Handler: admission.NewHandler(admission.Create, admission.Update),
+			}, nil
+		})
+}
+
+type customResourceDefinitionBound struct {
+	*admission.Handler
+}
+
+// Ensure that the required admission interfaces are implemented.
+var _ = admission.ValidationInterface(&customResourceDefinitionBound{})
+
+// Validate ensures that
+// - Has a valid type
+// - the the bound annotation exists that it is in the in the Shadow Workspace.
+func (o *customResourceDefinitionBound) Validate(ctx context.Context, a admission.Attributes, _ admission.ObjectInterfaces) (err error) {
+	if a.GetResource().GroupResource() != apiextensionsv1.Resource("customresourcedefinitions") {
+		return nil
+	}
+
+	if a.GetKind() != apiextensionsv1.Kind("CustomResourceDefinition").WithVersion("v1") {
+		return nil
+	}
+	crd, ok := a.GetObject().(*apiextensions.CustomResourceDefinition)
+	if !ok {
+		return fmt.Errorf("unexpected type %T", a.GetObject())
+	}
+
+	if _, ok := crd.GetAnnotations()[annotationKeyBoundCRD]; !ok {
+		// if no annotation pass through
+		return nil
+	}
+
+	name, err := request.ClusterNameFrom(ctx)
+	if err != nil {
+		return fmt.Errorf("failed to retrieve cluster from context: %w", err)
+	}
+	if name != apibinding.ShadowWorkspaceName {
+		return admission.NewForbidden(a, fmt.Errorf("unable to add crd with annotatio: %v in workspace: %v", annotationKeyBoundCRD, logicalcluster.From(crd)))
```

Comment:
typo

Reply:
```suggestion
		return admission.NewForbidden(a, fmt.Errorf("Annotation %q is reserved for system use", annotationKeyBoundCRD))
```

This got lost from a previous review + force-push

Reply:
Sorry!